### PR TITLE
OS#15992535 Revert #4652 to fix potential uint16 overflow in RegexParser

### PR DESCRIPTION
**Pull request by ricobbe**
My commit in PR #4652 introduced the possibility of an overflow in the regex parser: by changing RegexParser::numGroups and RegexParser::nextGroupId from int to uint16, we make it possible to overflow nextGroupId before hitting the check for an excessive number of capturing groups.

This PR is to remove this incorrect behavior before it propagates out of DEV3; I'll follow up with a later commit to adjust this API in a safer manner (and also possibly replace the AssertOrFailFast in RegexParser with something a little bit more user-friendly).

### CODE REVIEW

**Comment by ricobbe** [comment]
FYI, created #4722 for follow-up work.

File: lib/Runtime/Library/RegexHelper.cpp [hunk 1]
```diff
@@ -783,8 +786,9 @@ namespace Js
                     currentChar = replaceStr[substitutionOffset + 2];
                     if (currentChar >= _u('0') && currentChar <= _u('9'))
                     {
-                        uint16 tempCaptureIndex = (10 * captureIndex) + (uint16)(currentChar - _u('0'));
-                        Assert(tempCaptureIndex < 100); // numeric value of 2-digit positive decimal number
+                        // Should also be unsigned; see captureIndex above.
+                        int tempCaptureIndex = (10 * captureIndex) + (int)(currentChar - _u('0'));
+                        Assert(0 <= tempCaptureIndex && tempCaptureIndex < 100); // numeric value of 2-digit positive decimal number
```
Comment (Penguinwizzard) [inline lib/Runtime/Library/RegexHelper.cpp]: nit: The previous assert is 0<= x <= 9 and this is 0 <= x < 100; any chance of making both <=<= or <=<?

File: test/Regex/Bug15992535.baseline [hunk 2]
```diff
@@ -0,0 +1 @@
+Expecting fail fast
```
Comment (Penguinwizzard) [inline test/Regex/Bug15992535.baseline]: side note: long-term we should probably move this to throwing a catchable error instead, but this matches previous behavior.

Comment (ricobbe) [reply test/Regex/Bug15992535.baseline]: Yes, I agree; I've created #4722 for this follow-up work.

File: lib/Runtime/Library/RegexHelper.cpp (hunk 1 above)
Comment (ricobbe) [reply lib/Runtime/Library/RegexHelper.cpp]: Added note to follow-up task -- the hope with this change is to get it landed asap to reduce the chances of the broken code making it out of DEV3.

**Comment by ricobbe** [comment]
Note to reviewers: it does not seem possible to write a unit that is expected to fail fast and that works on all 3 platforms, at least without changing the test system.  Because the failfast behavior is temporary (see #4722), I've disabled the unit test on Linux & MacOS and will start work shortly on replacing the failfast with something better.

**Comment by ricobbe** [comment]
@dotnet-bot test Pre-Merge signing tool

**Pull request merged**